polyply/src/meta_molecule.py

```diff
@@ -360,3 +362,74 @@ def from_block(cls, force_field, mol_name):
         meta_mol = cls(graph, force_field=force_field, mol_name=mol_name)
         meta_mol.molecule = force_field.blocks[mol_name].to_molecule()
         return meta_mol
+
+    @classmethod
+    def from_cgsmiles_str(cls,force_field, cgsmiles_str, mol_name, seq_only=True, all_atom=False):
+        """
+        Constructs a :class::`MetaMolecule` from an CGSmiles string.
+        The force-field must contain the block with mol_name from
+        which to create the MetaMolecule. This function automatically
+        sets the MetaMolecule.molecule attribute.
+
+        Parameters
+        ----------
+        force_field: :class:`vermouth.forcefield.ForceField`
+            the force-field that must contain the block
+        cgsmiles_str:
+            the CGSmiles string describing the molecule graph
+        mol_name: str
+            name of the block matching a key in ForceField.blocks
+        seq_only: bool
+            if the string only describes the sequence; if this is False
+            then the molecule attribute is set
+        all_atom: bool
+            if the last molecule in the sequence is at all-atom resolution
+            can only be used if seq_only is False
+
+        Returns
+        -------
+        :class:`polyply.MetaMolecule`
+        """
+        if seq_only and all_atom:
+            msg = "You cannot define a sequence at all-atom level.\n"
+            raise IOError(msg)
+
+        # check if we have multiple resolutions
+        if cgsmiles_str.count('{') == 1:
+            meta_graph = read_cgsmiles(cgsmiles_str)
+            take_resname_from = 'fragname'
+        elif seq_only:
+            # initalize the cgsmiles molecule resolver
+            resolver = MoleculeResolver.from_string(cgsmiles_str, last_all_atom=all_atom)
+            # grep the last graph of the resolve iter
+            _, meta_graph = resolver.resolve_all()
+            take_resname_from = 'atomname'
+        else:
+            # initalize the cgsmiles molecule resolver
+            resolver = MoleculeResolver.from_string(cgsmiles_str, last_all_atom=all_atom)
+            *_, (meta_graph, molecule) = resolver.resolve_iter()
+            take_resname_from = 'fragname'
+
+        # we have to set some node attribute accoding to polyply specs
+        for node in meta_graph.nodes:
+            resname = meta_graph.nodes[node][take_resname_from]
+            meta_graph.nodes[node]['resname'] = resname
+            if not seq_only:
+                for atom in meta_graph.nodes[node]['graph'].nodes:
+                    meta_graph.nodes[node]['graph'].nodes[atom]['resname'] = resname
+                    meta_graph.nodes[node]['graph'].nodes[atom]['resname'] = resname
+                    molecule.nodes[atom]['resname'] = resname
+                    molecule.nodes[atom]['resid'] = node + 1
+                    #print(meta_graph.nodes[node]['graph'].nodes[atom])
+                    atomname = meta_graph.nodes[node]['graph'].nodes[atom]['atomname']
```

Comment:
While working on the effort in gh-390 to build SEBS, using this command (via `subprocess`): 

```
subprocess.CalledProcessError: Command '['polyply', 'gen_ff', '-i', 'smiles_molecule.acpype/smiles_molecule_GMX_OPLS.top', '-s', '{[#styrene]|2[#ethylene]|2[#butylene]|2[#styrene]|2}.{#styrene=[<]CC[>]c1ccccc1,#ethylene=[<]CCCC[>],#butylene=[<]CC[>](CC)}']' returned non-zero exit status 1.
```

I ended up with the traceback below the fold, affecting this line.

<details>

<summary> traceback </summary>

```
Traceback (most recent call last):
  File "/home/treddy/python_venvs/py_311_hemd/bin/polyply", line 292, in <module>
    main()
  File "/home/treddy/python_venvs/py_311_hemd/bin/polyply", line 288, in main
    subprogram(**args_dict)
  File "/home/treddy/python_venvs/py_311_hemd/lib/python3.11/site-packages/polyply/src/gen_ff.py", line 90, in gen_ff
    meta_mol = MetaMolecule.from_cgsmiles_str(force_field=force_field,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/treddy/python_venvs/py_311_hemd/lib/python3.11/site-packages/polyply/src/meta_molecule.py", line 424, in from_cgsmiles_str
    atomname = meta_graph.nodes[node]['graph'].nodes[atom]['atomname']
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: 'atomname'
```

</details>

The error was triggered by this CGSmiles string originally: `'{[#styrene]|2[#ethylene]|2[#butylene]|2[#styrene]|2}.{#styrene=[<]CC[>]c1ccccc1,#ethylene=[<]CCCC[>],#butylene=[<]CC[>](CC)}'`

I can reproduce the error with a much simpler CGSmiles string, `"{[#styrene]|2}.{#styrene=[<]CC[>]c1ccccc1}"`,
and even with: `'{[#styrene]}.{#styrene=[<]CC[>]c1ccccc1}'`.

Some debug printing shows the problematic atom in the context of the full CGSmiles input to be this one:

```
***** atom: 49
***** nodes: [45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56]
***** type(nodes): <class 'networkx.classes.reportviews.NodeView'>
***** atom index: {'charge': 0, 'aromatic': False, 'element': 'H', 'fragid': [3], 'fragname': 'ethylene', 'resname': 'ethylene'}
```

I'll see if I can dig a bit deeper..

Reply:
It turns out that `molecule.nodes[atom]['atomname']` is already populated in some cases, so the shim below seems to get past that particular issue:


<details>

```diff
--- a/polyply/src/meta_molecule.py
+++ b/polyply/src/meta_molecule.py
@@ -421,8 +421,12 @@ class MetaMolecule(nx.Graph):
                     molecule.nodes[atom]['resname'] = resname
                     molecule.nodes[atom]['resid'] = node + 1
                     #print(meta_graph.nodes[node]['graph'].nodes[atom])
-                    atomname = meta_graph.nodes[node]['graph'].nodes[atom]['atomname']
-                    molecule.nodes[atom]['atomname'] = atomname
+                    try:
+                        atomname = meta_graph.nodes[node]['graph'].nodes[atom]['atomname']
+                        molecule.nodes[atom]['atomname'] = atomname
+                    except KeyError:
+                        # the atomname tends to already be populated in some cases
+                        pass
```

</details>

But then there's yet another error:

<details>

```
Traceback (most recent call last):
  File "/home/treddy/python_venvs/py_311_hemd/lib/python3.11/site-packages/polyply/src/graph_utils.py", line 225, in find_one_ismags_match
    mapping = next(raw_matches)
              ^^^^^^^^^^^^^^^^^
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/treddy/python_venvs/py_311_hemd/bin/polyply", line 292, in <module>
    main()
  File "/home/treddy/python_venvs/py_311_hemd/bin/polyply", line 288, in main
    subprogram(**args_dict)
  File "/home/treddy/python_venvs/py_311_hemd/lib/python3.11/site-packages/polyply/src/gen_ff.py", line 97, in gen_ff
    unique_fragments, res_graph = FragmentFinder(target_mol).extract_unique_fragments(meta_mol.molecule)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/treddy/python_venvs/py_311_hemd/lib/python3.11/site-packages/polyply/src/fragment_finder.py", line 141, in extract_unique_fragments
    mapping = find_one_ismags_match(self.molecule,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/treddy/python_venvs/py_311_hemd/lib/python3.11/site-packages/polyply/src/graph_utils.py", line 228, in find_one_ismags_match
    raise IOError("no match_found")
OSError: no match_found
```

</details>

Reply:
@tylerjereddy I think this might be a dependency error, where the atomnames are different from what polyply expects. We've been changing quite a bit in the CGSmiles package over the past few weeks.I'll try taking a look to see if I can manage to figure it out.